text/0000-crt-static.md

```diff
@@ -0,0 +1,371 @@
+runtime [1], which require the entire 'world', possibly including std, to be
+compiled a certain way. C runtime linkage is another example of this
+requirement.
+
+[1]: https://internals.rust-lang.org/t/pre-rfc-a-vision-for-platform-architecture-configuration-specific-apis/3502
+
+From these observations we can design a cross-platform solution spanning both
+Cargo and the compiler by which Rust programs may link to either a dynamic or
+static C library, using only a single std binary. As future work this RFC
+discusses how the proposed scheme scheme can be extended to rebuild std
+specifically for a particular C-linkage scenario, which may have minor
+advantages on Windows due to issues around dllimport and dllexport; and how this
+scheme naturally extends to recompiling std in the presence of modified CPU
+features.
+
+This RFC does *not* propose unifying how the C runtime is linked across
+platforms (e.g. always dynamically or always statically) but instead leaves that
+decision to each target, and to future work.
+
+In summary the new mechanics are:
+
+- Specifying C runtime linkage via `-C target-feature=+crt-static` or `-C
+  target-feature=-crt-static`. This extends `-C target-feature` to mean not just
+  "CPU feature" ala LLVM, but "feature of the Rust target". Several existing
+  properties of this flag, the ability to add, with `+`, _or remove_, with `-`,
+  the feature, as well as the automatic lowering to `cfg` values, are crucial to
+  later aspects of the design. This target feature will be added to targets via
+  a small extension to the compiler's target specification.
+- Lowering `cfg` values to Cargo build script environment variables. This will
+  enable build scripts to understand all enabled features of a target (like
+  `crt-static` above) to, for example, compile C code correctly on MSVC.
+- Lazy link attributes. This feature is only required by std's own copy of the
+  libc crate, and only because std is distributed in binary form and it may yet
+  be a long time before Cargo itself can rebuild std.
+
+### Specifying dynamic/static C runtime linkage
+
+A new `target-feature` flag will now be supported by the compiler for relevant
+targets: `crt-static`. This can be enabled and disabled in the compiler via:
+
+```
+rustc -C target-feature=+crt-static ...
+rustc -C target-feature=-crt-static ...
+```
+
+Currently all `target-feature` flags are passed through straight to LLVM, but
+this proposes extending the meaning of `target-feature` to Rust-target-specific
+features as well. Target specifications will be able to indicate what custom
+target-features can be defined, and most existing targets will define a new
+`crt-static` feature which is turned off by default (except for musl).
+
+The default of `crt-static` will be different depending on the target. For
+example `x86_64-unknown-linux-musl` will have it on by default, whereas
+`arm-unknown-linux-musleabi` will have it turned off by default.
```

Comment:
To reiterate, we can't do that because it's a breaking change. There are many users relying on the fact that musl is a static target today. We can consider changing this all in the far future, but it is an explicitly stated non-goal of this RFC to attempt to perform any kind of unification of how the CRT is linked on various platforms.